app/updatemanager.go

```diff
@@ -150,22 +151,35 @@ type UpdateControlMapState struct {
 }
 
 type UpdateControlMap struct {
-	ID         string                           `json:"id"`
-	Priority   int                              `json:"priority"`
-	States     map[string]UpdateControlMapState `json:"states"`
-	expiryTime time.Time
-	expired    bool
+	ID                string                           `json:"id"`
+	Priority          int                              `json:"priority"`
+	States            map[string]UpdateControlMapState `json:"states"`
+	expiryTime        time.Time
+	expired           bool
+	mutex             sync.Mutex
+	expirationChannel chan struct{}
 }
 
 func (u *UpdateControlMap) Stamp(updateControlTimeoutSeconds int) *UpdateControlMap {
 	u.expiryTime = time.Now().Add(
 		time.Duration(updateControlTimeoutSeconds) * time.Second)
 	// expiration monitor
+	u.expire(time.Duration(updateControlTimeoutSeconds) * time.Second)
+	return u
+}
+
+func (u *UpdateControlMap) expire(in time.Duration) {
+	// TODO - Add a way to remove the timer upon object deletion from the pool
 	go func() {
-		<-time.After(time.Duration(updateControlTimeoutSeconds) * time.Second)
+		<-time.After(in)
+		u.mutex.Lock()
+		defer u.mutex.Unlock()
 		u.expired = true
```

Comment:
I haven't found any definitive answer online, but from what I can piece together, the lock triggers a "synchronization event" which flushes the processor cache and registers. So *hopefully* this should be safe. And mutexes would be pretty useless if it wasn't, so let's hope this common sense saves us! :-)